File: Scanner/src/ast/Statement.java
package ast;

import environment.Environment;

/**
 * Statement is a line of code which can be executed.
 * 
 * @version 3/16/18
 * @author David Melisso
 *
 */
public abstract class Statement
{
    /**
     * Executes the Statement
     * @param env
     *  the environment in which to execute the statement
     * @throws ASTException
     *  if there was an error executing the statement
     */
    public abstract void exec(Environment env) throws ASTException;
    
    /**
     * Throws a runtime exception because this method was not implemented in the child classes
     * @param e
     *  the Emitter which the child classes use to interface with the output file
     */
    public void compile(Emitter e)
    {
        throw new RuntimeException("Implement me!!!!!");
    }
}

File: JavaScriptInterpreter/ast/statement.js
/**
 * Expression stores values which it evaluates and returns the result, being either
 * an Integer, a String, or a Boolean
 *
 * @version 3/16/18
 * @author David Melisso
 *
 */
class Statement
{
    constructor() {
      if (new.target === Statement) {
        throw new TypeError("Cannot construct Statement instances directly");
      }

      if (typeof this.exec !== 'function') {
        throw new TypeError("Must override method exec");
      }
    }

    /**
     * https://stackoverflow.com/questions/29480569/does-ecmascript-6-have-a-convention-for-abstract-classes
     * Evaluates the expression and returns the result
     *
     * @param env
     *  the environment which to evaluate the expression in
     * @return
     *  the result of the expression
     * @throws ASTException
     *  if there was an error evaluating the expression
     */

}

module.exports = Statement;
